code/obj/machinery/turret.dm

```diff
@@ -379,74 +379,55 @@
-
-	if(!src.emagged)
-		if(src.locked && (!issilicon(user) && !isAI(user)))
-			t += "<I>(Swipe ID card to unlock control panel.)</I><BR>"
-		else
-			t += text("Turrets [] - <A href='?src=\ref[];toggleOn=1'>[]?</a><br><br>", src.enabled?"activated":"deactivated", src, src.enabled?"Disable":"Enable")
-			t += text("Currently set for [] - <A href='?src=\ref[];toggleLethal=1'>Change to []?</a><br><br>", src.lethal?"lethal":"stun repeatedly", src,  src.lethal?"Stun repeatedly":"Lethal")
-	else if(src.emagged)
-		var/o = ""
-		for(var/i=rand(4,50), i > 0, i--)
-			o += "kill[prob(50)?" ":null]"
-
-
-		for(var/i=1, i <= length(o), i++)
-			var/mod = rand(-5, 5)
-			t += text("<font size=[][]>[]</font>",mod>=0?"+":"-" ,mod , copytext(o, i, i+1))
-		t = "<B><font color=#FF0000>[t]</font></B>"
-		t += "<br><br>"
+	. = list(
+		"area" = istype(area) ? area.name : "Somewhere"
+	)
 
 
-	else
-		t += "!ALERT! Unable to connect to a turret!<br><br>"
-
-	user.Browse(t, "window=turretid")
-	onclose(user, "turretid")
-
-/obj/machinery/turretid/Topic(href, href_list)
-	..()
-	if ((!isliving(usr) && !isAIeye(usr)) || usr.getStatusDuration("stunned") || usr.getStatusDuration("weakened") || usr.stat)
+/obj/machinery/turretid/ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
+	if (..())
 		return
 	if (src.locked)
 		if (!issilicon(usr) && !isAI(usr))
 			boutput(usr, "Control panel is locked!")
 			return
 
-	if ((!issilicon(usr) && !isAIeye(usr)) && BOUNDS_DIST(usr, src) > 0)
-		return
-
-	if (href_list["toggleOn"])
-		src.enabled = !src.enabled
-		logTheThing("combat", usr, null, "turned [enabled ? "ON" : "OFF"] turrets from control \[[log_loc(src)]].")
-		src.updateTurrets()
-	else if (href_list["toggleLethal"])
-		src.lethal = !src.lethal
-		if(src.lethal)
-			logTheThing("combat", usr, null, "set turrets to LETHAL from control \[[log_loc(src)]].")
-			message_admins("[key_name(usr)] set turrets to LETHAL from control \[[log_loc(src)]].")
-		else
-			logTheThing("combat", usr, null, "set turrets to STUN from control \[[log_loc(src)]].")
-			message_admins("[key_name(usr)] set turrets to STUN from control \[[log_loc(src)]].")
-		src.updateTurrets()
-	src.Attackhand(usr)
+	switch (action)
+		if ("setEnabled")
+			src.enabled = params["enabled"]
+			logTheThing("combat", usr, null, "turned [enabled ? "ON" : "OFF"] turrets from control \[[log_loc(src)]].")
+			src.updateTurrets()
+		if ("setLethal")
+			src.lethal = params["lethal"]
+			if(src.lethal)
+				logTheThing("combat", usr, null, "set turrets to LETHAL from control \[[log_loc(src)]].")
+				message_admins("[key_name(usr)] set turrets to LETHAL from control \[[log_loc(src)]].")
+			else
+				logTheThing("combat", usr, null, "set turrets to STUN from control \[[log_loc(src)]].")
+				message_admins("[key_name(usr)] set turrets to STUN from control \[[log_loc(src)]].")
+			src.updateTurrets()
+	ui = tgui_process.try_update_ui(usr, src, ui)
```

Comment:
Have the relevant case statements return true instead (i.e. `. = TRUE`). This does the same thing under the hood but reduces unnecessary attempts to update.